### ISSUE TITLE
built-ins/Atomics/waitAsync/bigint/good-views.js & built-ins/Atomics/waitAsync/good-views.js are incorrect (timing dependent)

### ISSUE THREAD

**Issue by marjakh** [tc39/test262#2664]
In this test, all waitAsyncs resolve immediately; that doesn't create any timing issues. The worker does essentially this:

```
postMessage("message A");
postMessage("message B");
postMessage("message C");
```

The main thread does:

`Promise.all([getReportAsync(), getReportAsync(), getReportAsync()])`. When Promise.all resolves, it asserts that the resulting array is `["message A", "message B", "message C"]`.

Each `getReportAsync` creates a looper which checks if a message has been received, and if not, schedules itself to be re-executed with `setTimeout`: https://github.com/tc39/test262/blob/master/harness/atomicsHelper.js#L57

```
  $262.agent.getReportAsync = function() {
    return new Promise(function(resolve) {
      (function loop() {
        let result = getReport();
        if (!result) {
          setTimeout(loop, 1000);
        } else {
          resolve(result);
        }
      })();
    });
  };
```

Since the main thread does `getReportAsync` 3 times, the task queue will always have 3 tasks, related to 3 different `getReportAsync` calls. At some point, the foremost task will expire and post itself at the back of the queue etc.

Then a message is received. At that point, it's arbitrary which of the 3 loopers happens to own the topmost task of the task queue. That looper will get the message and stop looping.

As a result, the array returned by Promise.all will be an arbitrarily ordered collection of "message A", "message B" and "message C".

This is especially reproducible with d8, since d8 ignores the time parameter of `setTimeout`, spinning the loopers as fast as it can. With this setting, it's especially random which of the loopers happens to own the foremost task when a message arrives. With other js shells, it might not repro so easily; I'd assume the first 3 loops tasks would not find the message, but the next one (belonging to the first `getReportAsync`), executed 1 second after the time start, would.

Suggestions:
-> The test has to be fixed somehow. Probably would be best to separate it into 3 parts which are executed one at a time.
-> If this is how `getReportAsync` is implemented, test262 should allow only one `getReportAsync` to be active at any given time. This would prevent this error from happening again.

**Comment by marjakh** [comment]
Updated the title to reflect there are 2 of these tests.

**Comment by ljharb** [comment]
"ignores the time parameter of setTimeout" seems like an issue - doesn't that conflict with node's setTimeout impl as well as HTML's?

**Comment by marjakh** [comment]
This is only the d8 test runner, it's not used in Chrome or Node. d8 is not a standard-conforming js implementation, it's just implementing it "close enough" to run our tests.

So for example d8 does the wrong thing with this code:
setTimeout(foo, 10);
setTimeout(bar, 0);

(It runs foo first.)

With this test, it doesn't run the tasks in wrong order (since they're all posted with the same timeout), it just runs them earlier than the test intents.

Also, to be clear: d8 ignoring the setTimeout is *not* the reason why the test is wrong, it just makes this timing-dependent test more likely to fail.

**Comment by ljharb** [comment]
setTimeout isn't threading tho, since JS is single-threaded?

> ignoring the setTimeout is not the reason why the test is wrong,

Thanks for clarifying; I won't distract further with setTimeout chatter :-)

**Comment by marjakh** [comment]
The test creates a worker. The problem is that there's a race condition between the worker thread and the main thread. I tried to describe the race condition in detail in the OP, please let me know if more info is needed.

As d8 essentially "runs the main thread faster", the race condition becomes more obvious.

**Comment by marjakh** [comment]
cc @rwaldron as you added the tests

**Comment by rwaldron** [comment]
I'll take a look at these tomorrow. Thanks for the heads up.

**Comment by rwaldron** [comment]
Closing this. If there are any further problems, we can open new issues

**Comment by marjakh** [comment]
Has somebody actually fixed something? I didn't see any fixes. Please reopen or point me to a PR that fixes this.

**Comment by syg** [comment]
Reopening per https://github.com/tc39/test262/issues/2664#issuecomment-655668416. @rwaldron seems like nothing was acted on?

**Comment by leobalter** [comment]
I believe this might have been addressed at fcb3b91e4a from #2676.

The commit message had an inline "Fixes #2664", so it was not captured by GitHub.

**Comment by marjakh** [comment]
Hi, thanks, verified that that commit fixes the test. Sorry for the extra round trip :)
